// pdf-to-png-converter/out/pdf.to.png.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.pdfToPng = void 0;
const fs_1 = require("fs");
const path_1 = require("path");
const pdfjs = __importStar(require("pdfjs-dist/legacy/build/pdf"));
const const_1 = require("./const");
const node_canvas_factory_1 = require("./node.canvas.factory");
const props_to_pdf_doc_init_params_1 = require("./props.to.pdf.doc.init.params");
async function pdfToPng(pdfFilePathOrBuffer, props) {
    const isBuffer = Buffer.isBuffer(pdfFilePathOrBuffer);
    if (!isBuffer && !(0, fs_1.existsSync)(pdfFilePathOrBuffer)) {
        throw Error(`PDF file not found on: ${pdfFilePathOrBuffer}.`);
    }
    const pdfFileBuffer = isBuffer
        ? pdfFilePathOrBuffer
        : (0, fs_1.readFileSync)(pdfFilePathOrBuffer);
    const pdfDocInitParams = (0, props_to_pdf_doc_init_params_1.propsToPdfDocInitParams)(props);
    pdfDocInitParams.data = new Uint8Array(pdfFileBuffer);
    const pdfDocument = await pdfjs.getDocument(pdfDocInitParams).promise;
    const pngPagesOutput = [];
    const targetedPageNumbers = props?.pagesToProcess !== undefined
        ? props.pagesToProcess
        : Array.from({ length: pdfDocument.numPages }, (_, index) => index + 1);
    if (props?.strictPagesToProcess && targetedPageNumbers.some((pageNum) => pageNum < 1)) {
        throw new Error('Invalid pages requested, page number must be >= 1');
    }
    if (props?.strictPagesToProcess && targetedPageNumbers.some((pageNum) => pageNum > pdfDocument.numPages)) {
        throw new Error('Invalid pages requested, page number must be <= total pages');
    }
    if (props?.outputFolder && !(0, fs_1.existsSync)(props.outputFolder)) {
        (0, fs_1.mkdirSync)(props.outputFolder, { recursive: true });
    }
    for (const pageNumber of targetedPageNumbers) {
        if (pageNumber > pdfDocument.numPages || pageNumber < 1) {
            continue;
        }
        const page = await pdfDocument.getPage(pageNumber);
        const viewport = page.getViewport({
            scale: props?.viewportScale !== undefined
                ? props.viewportScale
                : const_1.PDF_TO_PNG_OPTIONS_DEFAULTS.viewportScale,
        });
        const canvasFactory = new node_canvas_factory_1.NodeCanvasFactory();
        const canvasAndContext = canvasFactory.create(viewport.width, viewport.height);
        const renderContext = {
            canvasContext: canvasAndContext.context,
            viewport,
            canvasFactory,
        };
        await page.render(renderContext).promise;
        let pageName;
        if (props?.outputFileMask) {
            pageName = props.outputFileMask;
        }
        if (!pageName && !isBuffer) {
            pageName = (0, path_1.parse)(pdfFilePathOrBuffer).name;
        }
        if (!pageName) {
            pageName = const_1.PDF_TO_PNG_OPTIONS_DEFAULTS.outputFileMask;
        }
        const pngPageOutput = {
            name: `${pageName}_page_${pageNumber}.png`,
            content: canvasAndContext.canvas.toBuffer(),
            path: '',
        };
        if (props?.outputFolder) {
            pngPageOutput.path = (0, path_1.resolve)(props.outputFolder, pngPageOutput.name);
            (0, fs_1.writeFileSync)(pngPageOutput.path, pngPageOutput.content);
        }
        pngPagesOutput.push(pngPageOutput);
    }
    return pngPagesOutput;
}
exports.pdfToPng = pdfToPng;
